engine/runtime/src/main/java/org/enso/interpreter/instrument/IdExecutionService.java

```diff
@@ -0,0 +1,264 @@
+        Object value,
+        String type,
+        String cachedType,
+        FunctionCallInfo callInfo,
+        FunctionCallInfo cachedCallInfo,
+        ProfilingInfo[] profilingInfo,
+        boolean wasCached) {
+      this.expressionId = expressionId;
+      this.value = value;
+      this.type = type;
+      this.cachedType = cachedType;
+      this.callInfo = callInfo;
+      this.cachedCallInfo = cachedCallInfo;
+      this.profilingInfo = profilingInfo;
+      this.wasCached = wasCached;
+    }
+
+    @Override
+    public String toString() {
+      String profilingInfo = Arrays.toString(this.profilingInfo);
+      return "ExpressionValue{"
+          + "expressionId="
+          + expressionId
+          + ", value="
+          + new MaskedString(value.toString()).applyMasking()
+          + ", type='"
+          + type
+          + '\''
+          + ", cachedType='"
+          + cachedType
+          + '\''
+          + ", callInfo="
+          + callInfo
+          + ", cachedCallInfo="
+          + cachedCallInfo
+          + ", profilingInfo="
+          + profilingInfo
+          + ", wasCached="
+          + wasCached
+          + '}';
+    }
+
+    /** @return the id of the expression computed. */
+    public UUID getExpressionId() {
+      return expressionId;
+    }
+
+    /** @return the type of the returned value. */
+    public String getType() {
+      return type;
+    }
+
+    /** @return the cached type of the value. */
+    public String getCachedType() {
+      return cachedType;
+    }
+
+    /** @return the computed value of the expression. */
+    public Object getValue() {
+      return value;
+    }
+
+    /** @return the function call data. */
+    public FunctionCallInfo getCallInfo() {
+      return callInfo;
+    }
+
+    /** @return the function call data previously associated with the expression. */
+    public FunctionCallInfo getCachedCallInfo() {
+      return cachedCallInfo;
+    }
+
+    /** @return the profiling information associated with this expression */
+    public ProfilingInfo[] getProfilingInfo() {
+      return profilingInfo;
+    }
+
+    /** @return whether or not the expression result was obtained from the cache */
+    public boolean wasCached() {
+      return wasCached;
+    }
+
+    /** @return {@code true} when the type differs from the cached value. */
+    public boolean isTypeChanged() {
+      return !Objects.equals(type, cachedType);
+    }
+
+    /** @return {@code true} when the function call differs from the cached value. */
+    public boolean isFunctionCallChanged() {
+      return !Objects.equals(callInfo, cachedCallInfo);
+    }
+  }
+
+  /** Information about the function call. */
+  public static class FunctionCallInfo {
+
+    private final QualifiedName moduleName;
+    private final QualifiedName typeName;
+    private final String functionName;
+
+    /**
+     * Creates a new instance of this class.
+     *
+     * @param call the function call.
+     */
```

Comment:
This is unrelated to this PR but I'm wondering if we should keep such comments as they do not really bring any new information.

Reply:
[f9cc871](https://github.com/enso-org/enso/pull/3509/commits/f9cc8711512c86f6f00f71b58734b6b6c4675e52)